mypy/plugins/attrs.py

```diff
@@ -1060,3 +1060,43 @@ def evolve_function_sig_callback(ctx: mypy.plugin.FunctionSigContext) -> Callabl
         fallback=ctx.default_signature.fallback,
         name=f"{ctx.default_signature.name} of {inst_type_str}",
     )
+
+
+def fields_function_sig_callback(ctx: mypy.plugin.FunctionSigContext) -> CallableType:
+    """Provide the proper signature for `attrs.fields`."""
+    if ctx.args and len(ctx.args) == 1 and ctx.args[0] and ctx.args[0][0]:
+        # <hack>
+        assert isinstance(ctx.api, TypeChecker)
+        inst_type = ctx.api.expr_checker.accept(ctx.args[0][0])
+        # </hack>
+        proper_type = get_proper_type(inst_type)
+
+        if isinstance(proper_type, AnyType):  # fields(Any) -> Any
+            return ctx.default_signature
+
+        cls = None
+        arg_types = ctx.default_signature.arg_types
+
+        if isinstance(proper_type, TypeVarType):
+            inner = get_proper_type(proper_type.upper_bound)
+            if isinstance(inner, Instance):
+                # We need to work arg_types to compensate for the attrs stubs.
+                arg_types = [inst_type]
```

Comment:
Curious what happens if we don't do this at all

Reply:
The test for generic classes starts failing if I remove it :(